File: examples/nim/crypto-lib/web3.nim
# Nim Web3 Client
# Systems programming for blockchain with expressive syntax

import httpclient, json, strutils, math

type
  Web3Client* = object
    rpcUrl*: string
    requestId*: int

proc newWeb3Client*(rpcUrl = "https://eth.llamarpc.com"): Web3Client =
  Web3Client(rpcUrl: rpcUrl, requestId: 0)

proc rpcCall(client: var Web3Client, `method`: string, params: JsonNode = newJArray()): JsonNode =
  client.requestId.inc

  let payload = %* {
    "jsonrpc": "2.0",
    "method": `method`,
    "params": params,
    "id": client.requestId
  }

  var httpClient = newHttpClient()
  httpClient.headers = newHttpHeaders({"Content-Type": "application/json"})

  let response = httpClient.postContent(client.rpcUrl, $payload)
  let jsonResponse = parseJson(response)

  if jsonResponse.hasKey("error"):
    raise newException(IOError, "RPC error: " & $jsonResponse["error"])

  return jsonResponse["result"]

proc hexToInt(hex: string): int64 =
  return parseBiggestInt(hex, 16)

proc weiToEther(wei: int64): float =
  return float(wei) / 1e18

proc weiToGwei(wei: int64): float =
  return float(wei) / 1e9

proc getBlockNumber*(client: var Web3Client): int64 =
  let result = client.rpcCall("eth_blockNumber")
  return hexToInt(result.getStr)

proc getBalance*(client: var Web3Client, address: string): float =
  let result = client.rpcCall("eth_getBalance", %* [address, "latest"])
  let wei = hexToInt(result.getStr)
  return weiToEther(wei)

proc getGasPrice*(client: var Web3Client): float =
  let result = client.rpcCall("eth_gasPrice")
  let wei = hexToInt(result.getStr)
  return weiToGwei(wei)

proc getTransaction*(client: var Web3Client, txHash: string): JsonNode =
  return client.rpcCall("eth_getTransactionByHash", %* [txHash])

when isMainModule:
  echo "👑 Nim Web3 Client"
  echo repeat('-', 40)

  var client = newWeb3Client()

  try:
    let blockNum = client.getBlockNumber()
    echo "📦 Block Number: ", blockNum

    let gas = client.getGasPrice()
    echo "⛽ Gas Price: ", formatFloat(gas, ffDecimal, 2), " Gwei"

    let address = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"
    let balance = client.getBalance(address)
    echo "💰 Balance: ", formatFloat(balance, ffDecimal, 4), " ETH"

  except IOError as e:
    echo "❌ Error: ", e.msg

File: examples/ocaml/type-safe-contracts/web3.ml
(* OCaml Web3 - Type-safe blockchain interactions *)

open Lwt.Infix

module Web3 = struct
  type t = { rpc_url: string; mutable request_id: int }

  let create ?(rpc_url = "https://eth.llamarpc.com") () =
    { rpc_url; request_id = 0 }

  let rpc_call client method_name params =
    client.request_id <- client.request_id + 1;

    let payload = `Assoc [
      ("jsonrpc", `String "2.0");
      ("method", `String method_name);
      ("params", `List params);
      ("id", `Int client.request_id)
    ] in

    let body = Yojson.Basic.to_string payload in

    Cohttp_lwt_unix.Client.post
      ~body:(Cohttp_lwt.Body.of_string body)
      ~headers:(Cohttp.Header.init_with "Content-Type" "application/json")
      (Uri.of_string client.rpc_url)
    >>= fun (resp, body) ->

    Cohttp_lwt.Body.to_string body
    >>= fun body_string ->

    let json = Yojson.Basic.from_string body_string in
    match json with
    | `Assoc assoc ->
        (match List.assoc_opt "result" assoc with
         | Some result -> Lwt.return result
         | None -> Lwt.fail_with "RPC error")
    | _ -> Lwt.fail_with "Invalid response"

  let hex_to_int = function
    | `String hex ->
        let hex = if String.sub hex 0 2 = "0x" then String.sub hex 2 (String.length hex - 2) else hex in
        int_of_string ("0x" ^ hex)
    | _ -> 0

  let wei_to_ether wei =
    float_of_int wei /. 1e18

  let wei_to_gwei wei =
    float_of_int wei /. 1e9

  let get_block_number client =
    rpc_call client "eth_blockNumber" []
    >|= hex_to_int

  let get_balance client address =
    rpc_call client "eth_getBalance" [`String address; `String "latest"]
    >|= hex_to_int
    >|= wei_to_ether

  let get_gas_price client =
    rpc_call client "eth_gasPrice" []
    >|= hex_to_int
    >|= wei_to_gwei
end

(* Example usage *)
let () =
  Lwt_main.run begin
    let client = Web3.create () in
    Printf.printf "🐫 OCaml Web3 Client\n%!";
    Printf.printf "%s\n%!" (String.make 40 '-');

    Web3.get_block_number client
    >>= fun block ->
    Printf.printf "📦 Block Number: %d\n%!" block;

    Web3.get_gas_price client
    >>= fun gas ->
    Printf.printf "⛽ Gas Price: %.2f Gwei\n%!" gas;

    let address = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045" in
    Web3.get_balance client address
    >|= fun balance ->
    Printf.printf "💰 Balance: %.4f ETH\n%!" balance
  end
